### PR TITLE
Refactor SharedDriver lifecycle in FluentAdapter (JUnit, TestNG & Cucumber)

### LINKED ISSUE #196: SharedDriver with SharedType.ONCE is not destroyed on shutdown
When I create a .@SharedDriver with SharedType.ONCE - i.e. the default - it is not destroyed on shutdown.
This is because of a bug in the shutdown hook. The shutdown hook tries to find a thread local driver, which is obviously empty since the hook  runs in a different thread than the tests.
It should instead close the driver stored in the sharedDriver field.

### PR DESCRIPTION
**Pull request by Toilal**
It should fix #196 properly, and should also fix your ghost browser problem @filipcynarski 

Basically, the idea is to use a Singleton responsible of the lifecycle of WebDriver created by FluentTest / FluentTestNG. This singleton provides methods to get or create the driver for each test method, using the ```testClass```, ```testName``` and value of ```@SharedDriver``` strategy. 

It also has methods to release/quit drivers when the test method is finished, the test class is finished, and register the Shutdown hooks to release all running WebDriver when the JVM shutdowns.

### CODE REVIEW

File: fluentlenium-core/src/main/java/org/fluentlenium/adapter/FluentTestRule.java [hunk 1]
```diff
@@ -25,6 +25,8 @@ public void evaluate() throws Throwable {
                 } catch (Throwable e) {
                     failed(e, description);
                     throw e;
+                } finally {
+                    finished(description);
```
Comment (Toilal) [inline fluentlenium-core/src/main/java/org/fluentlenium/adapter/FluentTestRule.java]: This one is missing from ```master``` branch too !

File: fluentlenium-core/src/test/java/org/fluentlenium/adapter/SharedWebDriverContainerTest.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    public void getOrCreateDriver_with_different_test_names_creates_distinct_instances() {
+        SharedWebDriver driver = container.getOrCreateDriver(this, Object.class, "test", SharedDriverStrategy.PER_METHOD);
+        SharedWebDriver driver2 = container.getOrCreateDriver(this, Object.class, "otherTest", SharedDriverStrategy.PER_METHOD);
+
+        assertThat(driver).isNotEqualTo(driver2);
+        assertThat(container.getAllDrivers()).containsOnly(driver, driver2);
+        assertThat(container.getTestClassDrivers(Object.class)).containsOnly(driver, driver2);
+
+        container.quit(driver);
+        container.quit(driver2);
```
Comment (filipcynarski) [inline fluentlenium-core/src/test/java/org/fluentlenium/adapter/SharedWebDriverContainerTest.java]: I'd add a check is only one driver left in a middle

File: fluentlenium-core/src/test/java/org/fluentlenium/adapter/SharedWebDriverContainerTest.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        SharedWebDriver driver = container.getOrCreateDriver(this, Object.class, "test", SharedDriverStrategy.PER_METHOD);
+        SharedWebDriver driver2 = container.getOrCreateDriver(this, String.class, "test", SharedDriverStrategy.PER_METHOD);
+
+        assertThat(driver).isNotEqualTo(driver2);
+        assertThat(container.getAllDrivers()).containsOnly(driver, driver2);
+        assertThat(container.getTestClassDrivers(Object.class)).containsOnly(driver);
+        assertThat(container.getTestClassDrivers(String.class)).containsOnly(driver2);
+
+        container.quit(driver);
+        container.quit(driver2);
```
Comment (filipcynarski) [inline fluentlenium-core/src/test/java/org/fluentlenium/adapter/SharedWebDriverContainerTest.java]: the same semi assert would be helpful

File: fluentlenium-core/src/test/java/org/fluentlenium/adapter/SharedWebDriverContainerTest.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        SharedWebDriver driver = container.getOrCreateDriver(this, Object.class, "test", SharedDriverStrategy.PER_CLASS);
+        SharedWebDriver driver2 = container.getOrCreateDriver(this, String.class, "otherTest", SharedDriverStrategy.PER_CLASS);
+
+        assertThat(driver).isNotEqualTo(driver2);
+        assertThat(container.getAllDrivers()).containsOnly(driver, driver2);
+        assertThat(container.getTestClassDrivers(Object.class)).containsOnly(driver);
+        assertThat(container.getTestClassDrivers(String.class)).containsOnly(driver2);
+
+        container.quit(driver);
+        container.quit(driver2);
```
Comment (filipcynarski) [inline fluentlenium-core/src/test/java/org/fluentlenium/adapter/SharedWebDriverContainerTest.java]: I'd add assert in a middle is driver2 still active before quit

**Comment by filipcynarski** [comment]
I'm merging this one for 1.0.0-SNAPSHOT

**Pull request merged**